cdc/model/codec/codec.go

```diff
@@ -0,0 +1,270 @@
+		switch v := redoC.Value.(type) {
+		case []byte:
+			if len(v) == 0 {
+				redoC.ValueIsEmptyByteSlice = true
+			}
+		}
+	}
+
+	if r.RedoRow.Row != nil {
+		row := r.RedoRow.Row
+		r.RedoRow.Columns = make([]model.RedoColumn, 0, len(row.Columns))
+		r.RedoRow.PreColumns = make([]model.RedoColumn, 0, len(row.PreColumns))
+		for _, c := range row.Columns {
+			redoC := model.RedoColumn{}
+			if c != nil {
+				redoC.Value = c.Value
+				redoC.Flag = uint64(c.Flag)
+				workaroundColumn(&redoC)
+			}
+			r.RedoRow.Columns = append(r.RedoRow.Columns, redoC)
+		}
+		for _, c := range row.PreColumns {
+			redoC := model.RedoColumn{}
+			if c != nil {
+				redoC.Value = c.Value
+				redoC.Flag = uint64(c.Flag)
+				workaroundColumn(&redoC)
+			}
+			r.RedoRow.PreColumns = append(r.RedoRow.PreColumns, redoC)
+		}
+	}
+	if r.RedoDDL.DDL != nil {
+		r.RedoDDL.Type = byte(r.RedoDDL.DDL.Type)
+	}
+}
+
+func DecodeRedoLog(dc *msgp.Reader) (r *model.RedoLog, err error) {
+	var p []byte
+	p, err = dc.R.Peek(versionPrefixLength)
+	if err != nil {
+		return
+	}
+
+	shouldBeV1 := false
+	for i := 0; i < versionPrefixLength; i++ {
+		if p[i] != versionPrefix[i] {
+			shouldBeV1 = true
+			break
+		}
+	}
+
+	if shouldBeV1 {
+		var rv1 *codecv1.RedoLog = new(codecv1.RedoLog)
+		if err = rv1.DecodeMsg(dc); err != nil {
+			return
+		}
+		codecv1.PostUnmarshal(rv1)
+		r = redoLogFromV1(rv1)
+	} else {
+		p = make([]byte, 4)
+		if _, err = dc.R.Read(p); err != nil {
+			return
+		}
+		version, _ := decodeVersion(p[versionPrefixLength:])
+		if version == latestVersion {
+			r = new(model.RedoLog)
+			if err = r.DecodeMsg(dc); err != nil {
+				return
+			}
+			postUnmarshal(r)
+		} else {
+			panic("unsupported codec version")
+		}
+	}
+	return
+}
+
+func EncodeRedoLog(r *model.RedoLog, en *msgp.Writer) (err error) {
+	preMarshal(r)
+	if _, err = en.Write(versionPrefix[:]); err != nil {
+		return
+	}
+	versionField := binary.BigEndian.AppendUint16(nil, latestVersion)
+	if _, err = en.Write(versionField); err != nil {
+		return
+	}
+	return r.EncodeMsg(en)
+}
+
+func UnmarshalRedoLog(bts []byte) (r *model.RedoLog, o []byte, err error) {
+	if len(bts) < versionPrefixLength {
+		err = msgp.ErrShortBytes
+		return
+	}
+
+	shouldBeV1 := false
+	for i := 0; i < versionPrefixLength; i++ {
+		if bts[i] != versionPrefix[i] {
+			shouldBeV1 = true
+			break
+		}
+	}
+	if shouldBeV1 {
+		var rv1 *codecv1.RedoLog = new(codecv1.RedoLog)
+		if o, err = rv1.UnmarshalMsg(bts); err != nil {
+			return
+		}
+		codecv1.PostUnmarshal(rv1)
+		r = redoLogFromV1(rv1)
+	} else {
+		bts = bts[versionPrefixLength:]
+		version, bts := decodeVersion(bts)
+		if version == latestVersion {
+			r = new(model.RedoLog)
+			if o, err = r.UnmarshalMsg(bts); err != nil {
+				return
+			}
+			postUnmarshal(r)
+		} else {
+			panic("unsupported codec version")
+		}
+	}
+	return
+}
+
+func MarshalRedoLog(r *model.RedoLog, b []byte) (o []byte, err error) {
+	preMarshal(r)
```

Comment:
`preMarshal` was a little confused. Maybe we can encapsulate such logic as `ToRedoLog` method?

Ref: https://github.com/pingcap/tiflow/pull/8107/files#diff-eb40c4235fa377719c71b4d416eb38d4f4134b281758dd43687ca525d0ca3549R272

Reply:
I don't want to expose interface `ToRedoLog`. All logic in the function is for marshaling. Maybe we can add a function `MarshalRowAsRedoLog`.

Reply:
Done. PTAL again, thanks!